package_modules/apt_get

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/python
+
+import sys
+import os
+import subprocess
+import re
+
+dpkg_cmd = "/usr/bin/dpkg"
+dpkg_deb_cmd = "/usr/bin/dpkg-deb"
+dpkg_query_cmd = "/usr/bin/dpkg-query"
+dpkg_output_format = "Name=${Package}\nVersion=${Version}\nArchitecture=${Architecture}\n"
+dpkg_status_format = "Status=${Status}\n" + dpkg_output_format
+
+apt_get_options = ["-o", "Dpkg::Options::=--force-confold", "-o", "Dpkg::Options::=--force-confdef", "-y", "--force-yes"]
+apt_get_cmd = "/usr/bin/apt-get"
+
+os.environ['DEBIAN_FRONTEND'] = "noninteractive"
+os.environ['LC_ALL'] = "C"
+
+if (os.environ.get('CFENGINE_TEST_DPKG_CMD')):
+    dpkg_cmd = os.environ['CFENGINE_TEST_DPKG_CMD']
+if (os.environ.get('CFENGINE_TEST_DPKG_DEB_CMD')):
+    dpkg_deb_cmd = os.environ['CFENGINE_TEST_DPKG_DEB_CMD']
+if (os.environ.get('CFENGINE_TEST_DPKG_QUERY_CMD')):
+    dpkg_query_cmd = os.environ['CFENGINE_TEST_DPKG_QUERY_CMD']
+if (os.environ.get('CFENGINE_TEST_APT_GET_CMD')):
+    apt_get_cmd = os.environ['CFENGINE_TEST_APT_GET_CMD']
+
+def get_package_data():
+    pkg_string = ""
+    for line in sys.stdin:
+        result = re.match("^File=(.*)", line)
+        if (result):
+            pkg_string = result.group(1)
+            break
+
+    if (pkg_string == ""):
+        return 1
+
+    repo = False
+    if (re.match("^/.*", pkg_string)):
+        # Absolute file.
+        sys.stdout.write("PackageType=file\n")
+    elif (re.search("([:,]|-[0-9])", pkg_string)):
+        # Contains either a version number or an illegal symbol.
+        sys.stdout.write(line + "ErrorMessage: Package string with illegal format\n")
+        return 1
+    else:
+        repo = True
+        sys.stdout.write("PackageType=repo\n")
+
+    if (repo):
+        sys.stdout.write("Name=" + pkg_string + "\n")
+        return 0
+    else:
+        sys.stdout.flush()
+        return subprocess.call([dpkg_deb_cmd, "--showformat", dpkg_output_format, "-W", result.group(1)])
+
+
+def list_installed():
+    for line in sys.stdin:
+        # Ignore everything.
+        pass
+
+    process = subprocess.Popen([dpkg_query_cmd, "--showformat", dpkg_status_format, "-W"], stdout=subprocess.PIPE)
+    installed_package = False
+    for line in process.stdout:
+        if line[-1] == "\n":
+            line = line[:-1]
+        # The space before "installed" is important, because it can be "not-installed".
+        if re.match("^Status=.* installed.*", line):
+            installed_package = True
+        elif re.match("^Status=.*", line):
+            installed_package = False
```

Comment:
I think handling offsets is both more error prone and less clear than letting the regex handle it. Speed is not an issue here, since the package manager will have execution time orders of magnitude higher than the script.

The second one can use ``startswith()`` though, since it is 100% equivalent. Fixed that.